Havtorn/Source/Engine/Core/MathTypes/Matrix.h

```diff
@@ -727,29 +732,53 @@ namespace Havtorn
 		translation = matrix.GetTranslation();	
 	}
 
+	inline void SMatrix::Decompose(const SMatrix& matrix, F32* translationData, F32* rotationData, F32* scaleData)
+	{
+		const SVector& t = matrix.GetTranslation();
+		translationData[0] = t.X;
+		translationData[1] = t.Y;
+		translationData[2] = t.Z;
+
+		const SVector& r = matrix.GetEuler();
+		rotationData[0] = r.X;
+		rotationData[1] = r.Y;
+		rotationData[2] = r.Z;
+
+		const SVector& s = matrix.GetScale();
+		scaleData[0] = s.X;
+		scaleData[1] = s.Y;
+		scaleData[2] = s.Z;
+	}
+	
 	inline void SMatrix::Recompose(const SVector& translation, const SVector& rotation, const SVector& scale, SMatrix& outMatrix)
 	{
-		SMatrix rot[3];
-		SVector unaryDirections[3] = { SVector::Right, SVector::Up, SVector::Forward };
-		F32 rotations[3] = { rotation.X, rotation.Y, rotation.Z };
-		for (U8 i = 0; i < 3; i++)
-		{
-			rot[i] = SMatrix::CreateRotationAroundAxis(UMath::DegToRad(rotations[i]), unaryDirections[i]);
-		}
+		F32 t[3] = { translation.X, translation.Y, translation.Z };
+		F32 r[3] = { rotation.X, rotation.Y, rotation.Z };
+		F32 s[3] = { scale.X, scale.Y, scale.Z };
+		Recompose(t, r, s, outMatrix);
+	}
+
+	inline void SMatrix::Recompose(const F32* translationData, const F32* rotationData, const F32* scaleData, SMatrix& outMatrix) 
+	{
+		HV_ASSERT((translationData), "Passing nullptr [translationData] to Recompose!");
+		HV_ASSERT((rotationData), "Passing nullptr [rotationData] to Recompose!");
+		HV_ASSERT((scaleData), "Passing nullptr [scaleData] to Recompose!");
 
-		outMatrix = rot[2] * rot[1] * rot[0];
+		// AG: Rotation Order XYZ used to not cause issues with ImGuizmo used in InspectorWindow.
+		outMatrix = CreateRotationAroundX(UMath::DegToRad(rotationData[0]));
+		outMatrix *= CreateRotationAroundY(UMath::DegToRad(rotationData[1]));
+		outMatrix *= CreateRotationAroundZ(UMath::DegToRad(rotationData[2]));
 
-		F32 scales[3] = { scale.X, scale.Y, scale.Z };
 		F32 validScale[3] = { 1.0f, 1.0f, 1.0f };
 		for (U8 i = 0; i < 3; i++)
 		{
-			if (UMath::FAbs(scales[i]) < FLT_EPSILON)
+			if (UMath::FAbs(scaleData[i]) < FLT_EPSILON)
 			{
 				validScale[i] = 0.001f;
```

Comment:
could probably replace this bad boi with a define of some sort, check SMALL_NUMBER for ref